### PR TITLE
Cache split in TPE for high-dimensional optimization

### PR DESCRIPTION
**Pull request by nabenabe0928**
## Motivation

As TPE significantly slows down for high-dimensional optimization, this PR introduces caching mechanism for the TPE split.

## Description of the changes

- Cache the split information in sampler so that we can re-use the result

For `n_trials=1000` and `dim=10`, the runtimes are the following:

|This PR |Master|
|:--:|:--:|
|29|47|

### CODE REVIEW

**Comment by not522** [comment]
What is the relationship between this PR and #5454? Should we review it after #5454?

**Comment by nabenabe0928** [comment]
@not522 
These two PRs are orthogonal works, so we can separately work!
This PR aims to share the split information in a trial.
The other PR aims to share the information over multiple trials giving the same set of arguments to the hssp solver.

**Comment by not522** [comment]
@eukaryo @gen740 Could you review this PR?

File: optuna/_hypervolume/utils.py [hunk 1]
```diff
@@ -1,3 +1,5 @@
+from __future__ import annotations
```
Comment (gen740) [inline optuna/_hypervolume/utils.py:1]: I believe this change is not relevant to this PR.
This change would be good to include when using features from ﻿`__future__`.

**Comment by eukaryo** [comment]
Sorry, I am temporarily busy because my computer is not working, and I suppose @HideakiImamura -san is the appropriate reviewer. @HideakiImamura, could you review this PR?

File: optuna/_hypervolume/utils.py (hunk 1 above)
Comment (nabenabe0928) [reply optuna/_hypervolume/utils.py]: Done!

**Review by HideakiImamura — COMMENTED** [review]
Holding cache on the sampler instance causes the data inconsistency across different processes. Could you give me an opinion that?

**Comment by nabenabe0928** [comment]
We discussed internally and decided to close this PR.

**Comment by nabenabe0928** [comment]
Just for future reminder, I will leave some comments:

### This PR does not cause any issues between processes

As each trial is sampled in a thread and a sampling of a specific trial will not be scattered on multiple processes or threads, we do not have to be concerned about the missing cached data in another  process.

However, when using multiple threads, we need to take care of another thread overwriting the cache before one trial completes its sampling.
For this reason, I added a buffer to store the split data up to the latest 64 trials in a thread.

Anyways, missing cache does not cause any issue because we simply need to re-calculate the split, which incurs some more computational burden.

**Pull request closed without merging**